:chart_with_upwards_trend: 🧪 BN254 + `modexp` Precompile Circuits and Tests

ZamDimon opened the pull request:
# What ❔

This pull request adds circuits for three EC operations over BN254 elliptic curve $E$:
- `ecadd`: adding two point on $E(\mathbb{F}_p)$;
- `ecmul`: multiplying point on $E(\mathbb{F}_p)$ by a scalar in $\mathbb{F}_q$. 
- `ecpairing`: find pairing between the point on the base curve $E(\mathbb{F}_p)$ and twisted one on $E'(\mathbb{F}\_{p^2})$

We also include the `modexp` precompile circuit.

Besides circuit implementations, this pull request contains a comprehensive list of tests to validate that all three operations + `modexp` are working correctly. This is done via generating tests in the JSON format via `SageMath` code, which works quite well with Mathematics.

## Why ❔

This has not been implemented, thus the pull request.

## Checklist

<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

ZamDimon committed:
:tada: add basic setup for `bn254` circuits

ZamDimon committed:
:tada: add basic setup for `bn254` circuits

ZamDimon committed:
:construction: add tests for `ec_mul`, trying to make them work

ZamDimon committed:
:construction: add tests for `ec_mul`, trying to make them work

ZamDimon committed:
:test_tube: fix expected values in scalar decomposition

ZamDimon committed:
:test_tube: fix expected values in scalar decomposition

ZamDimon committed:
:test_tube: tests work

ZamDimon committed:
:test_tube: tests work

ZamDimon committed:
:test_tube: add `ecadd` test

ZamDimon committed:
:test_tube: add `ecadd` test

ZamDimon committed:
:test_tube: :sparkles: add more convenient test structure

ZamDimon committed:
:test_tube: :sparkles: add more convenient test structure

ZamDimon committed:
:test_tube: add even better tests via json files

ZamDimon committed:
:test_tube: add even better tests via json files

ZamDimon committed:
:test_tube: add `fq2` test

ZamDimon committed:
:test_tube: add `fq2` test

ZamDimon committed:
:test_tube: :adhesive_bandage: fully fix and finalize fq6 field extension tests

ZamDimon committed:
:test_tube: :adhesive_bandage: fully fix and finalize fq6 field extension tests

ZamDimon committed:
:memo: add documentation for `bn254` folder

ZamDimon committed:
:memo: add documentation for `bn254` folder

ZamDimon committed:
:test_tube: add `g2` curve tests

ZamDimon committed:
:test_tube: add `g2` curve tests

ZamDimon committed:
:technologist: make final exponentiation work

ZamDimon committed:
:technologist: make final exponentiation work

ZamDimon committed:
:memo: make final exponentiation comments a bit neater

ZamDimon committed:
:memo: make final exponentiation comments a bit neater

ZamDimon committed:
:sparkles: add more `ecpairing` tests

ZamDimon committed:
:sparkles: add more `ecpairing` tests

ZamDimon committed:
:test_tube: further `ecpairing` struggling with tests

ZamDimon committed:
:test_tube: further `ecpairing` struggling with tests

ZamDimon committed:
:tada: `ecpairing` in sage works

ZamDimon committed:
:tada: `ecpairing` in sage works

ZamDimon committed:
:fire: making `ecpairing` sage file shorter

ZamDimon committed:
:fire: making `ecpairing` sage file shorter

ZamDimon committed:
:art: small polishes

ZamDimon committed:
:art: small polishes

nikitamasych committed:
:sparkles: ecmul entrypoint

nikitamasych committed:
:sparkles: ecmul entrypoint

nikitamasych committed:
:sparkles: ecadd entrypoint

nikitamasych committed:
:sparkles: ecadd entrypoint

nikitamasych committed:
:sparkles: ecpairing entrypoint

nikitamasych committed:
:sparkles: ecpairing entrypoint

ZamDimon committed:
:test_tube: make tests smaller

ZamDimon committed:
:test_tube: make tests smaller

ZamDimon committed:
:truck: transfer `bn254`-specific functions to the repository

ZamDimon committed:
:truck: transfer `bn254`-specific functions to the repository

nikitamasych committed:
:adhesive_bandage: post-review fixes

nikitamasych committed:
:adhesive_bandage: post-review fixes

nikitamasych committed:
:zap: remove redundant normalizations in pairing

nikitamasych committed:
:zap: remove redundant normalizations in pairing

nikitamasych committed:
Merge branch 'v1.5.0' into feature/bn254-circuits

jules commented on src/bn254/ec_add/mod.rs:
```diff
@@ -0,0 +1,330 @@
+) -> (Boolean<F>, UInt256<F>, UInt256<F>) {
+    let base_field_params = &Arc::new(bn254_base_field_params());
+
+    // TODO: add validation of input values
+    let x1 = convert_uint256_to_field_element(cs, &x1, base_field_params);
+    let y1 = convert_uint256_to_field_element(cs, &y1, base_field_params);
+    let mut point1 = BN256SWProjectivePoint::from_xy_unchecked(cs, x1, y1);
+
+    let x2 = convert_uint256_to_field_element(cs, &x2, base_field_params);
+    let y2 = convert_uint256_to_field_element(cs, &y2, base_field_params);
```
i think this can be checked with 4 subtractions from modulus right?

nikitamasych replied:
Do you mean validation of inputs? I can't quite get how it will work 👨‍💻

jules replied:
oh, my bad - yeah it would be a bit more involved. i'm assuming the TODO means you want to ensure that x and y make a valid curve point

nikitamasych replied:
yeah, it's just that in yul contracts we do have ready validation that the point is on curve, so this will be duplicating with worse performance; I would though wait for the review from era-contracts and see what guys would say - shall we leverage the checks here or no. Plus further integration I assume will tell a bit more on this subject

jules replied:
makes sense - in this case doing pre validation in the contracts would be cheaper for sure

jules commented on src/bn254/ec_mul/mod.rs:
```diff
@@ -0,0 +1,329 @@
+
+    // TODO: add validation of input values
+    let x = convert_uint256_to_field_element(cs, &x, base_field_params);
+    let y = convert_uint256_to_field_element(cs, &y, base_field_params);
```
same here, just subtractions with modulus should fix it right?

jules reviewed (commented):
only tiny suggestions - all entry points look great atm and tests seem to work fine!

nikitamasych committed:
:recycle: removed todos for validation in ecadd, ecmul -> vanguard handling in yul

nikitamasych committed:
:recycle: removed todos for validation in ecadd, ecmul -> vanguard handling in yul

nikitamasych committed:
:building_construction: changed ecpairing to work with multiplies of G1 and G2

nikitamasych committed:
:building_construction: changed ecpairing to work with multiplies of G1 and G2

nikitamasych committed:
:safety_vest: added field validation for input coordinates in pairing

nikitamasych committed:
:safety_vest: added field validation for input coordinates in pairing

nikitamasych committed:
:ambulance: corrected accumulated pairing

nikitamasych committed:
:ambulance: corrected accumulated pairing

nikitamasych committed:
:bug: removed need for copy trait in ecpairing input

nikitamasych committed:
:bug: removed need for copy trait in ecpairing input

ZamDimon committed:
:sparkles: add base layer circuit types for EC operations and modexp (#1)

ZamDimon committed:
:sparkles: add base layer circuit types for EC operations and modexp (#1)

ZamDimon committed:
:sparkles: :test_tube: add torus tests

nikitamasych committed:
:safety_vest: added base field validation for BN254 ecadd, ecmul and ecpairing

ZamDimon committed:
:adhesive_bandage: fix comments in torus, create a separate function with torus implementation

nikitamasych committed:
:safety_vest: completed validation for ecadd and ecmul

nikitamasych committed:
:safety_vest: completed validation for ecpairing

jules commented on Cargo.toml:
```diff
@@ -15,16 +15,24 @@ smallvec = { version = "*", features = [
 # zkevm_opcode_defs = {path = "../era-zkevm_opcode_defs"}
+
 cs_derive = { git = "https://github.com/matter-labs/era-boojum.git", branch = "main" }
-boojum = {git = "https://github.com/matter-labs/era-boojum.git", branch = "main"}
```
we should point back to the canon boojum, specifically `dl-precompiles` once it is merged in

jules replied:
i know you can't work with it until it is merged but just flagging in case we may forget :)

jules commented on src/bn254/ec_add/implementation.rs:
```diff
@@ -0,0 +1,16 @@
+
+use super::*;
+
+/// Adds two points in the plain SW projective coordinates.
+pub fn projective_add<F, CS>(
+    cs: &mut CS,
+    point_1: &mut BN256SWProjectivePoint<F>,
+    mut point_2: (BN256BaseNNField<F>, BN256BaseNNField<F>),
+) -> BN256SWProjectivePoint<F>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    point_1.add_mixed(cs, &mut point_2)
+}
```
this doesnt' actually seem to be used in-circuit and looks like just a function wrapper so i think we can delete it

nikitamasych replied:
It is used for testing and I can rewrite the entrypoint for ecadd to utilize it :) It creates homogeneous codebase

jules commented on src/bn254/ec_add/mod.rs:
```diff
@@ -0,0 +1,361 @@
+
+    let precompile_address = UInt160::allocated_constant(
+        cs,
+        *zkevm_opcode_defs::system_params::ECRECOVER_INNER_FUNCTION_PRECOMPILE_FORMAL_ADDRESS, // TODO: change to ECADD_PRECOMPILE_FORMAL_ADDRESS
```
did we already decide on a formal address here?

nikitamasych replied:
We have forked `zkevm_opcode_defs` and updated similarly there:

https://github.com/distributed-lab/era-zkevm_opcode_defs/blob/feature/ec-and-modexp-precompiles/src/system_params.rs#L32

jules commented on src/bn254/ec_pairing/implementation.rs:
```diff
@@ -0,0 +1,512 @@
+    /// The final exponentiation is partially based on _Algorithm 31_ from
+    /// https://eprint.iacr.org/2010/354.pdf, but mainly based on implementation
+    /// from pairing repository https://github.com/matter-labs/pairing.
+    pub fn evaluate(cs: &mut CS, r: &mut BN256Fq12NNField<F>) -> Self {
```
was it still necessary to have this many normalization calls?

nikitamasych replied:
We reduced the amount in overall implementation, but for `FinalExpEvaluation` evaluate there is no much space for it

jules commented on src/bn254/validation.rs:
```diff
@@ -0,0 +1,137 @@
+
+    let x_c0_is_zero = x_c0.is_zero(cs);
+    let x_c1_is_zero = x_c1.is_zero(cs);
+    let x_is_zero = x_c0_is_zero.and(cs, x_c1_is_zero);
```
i think the single `and` calls can be removed and all 4 booleans can be put in the multi and gate, it should be a touch cleaner

jules commented on src/bn254/validation.rs:
```diff
@@ -0,0 +1,137 @@
+    let x_is_zero = x.is_zero(cs);
+    let y_is_zero = y.is_zero(cs);
+
+    Boolean::multi_and(cs, &[x_is_zero, y_is_zero])
```
can just use a normal and gate here

jules reviewed (changes requested):
another couple points but coming together very nicely now!

nikitamasych committed:
:art: fixes after review

nikitamasych committed:
:sparkles: added ecadd, ecmul, ecpairing to scheduler

nikitamasych committed:
:rotating_light: cargo fix & cargo fmt

nikitamasych commented:
Rebased with respect to `dl-precompiles` 👀 Were 4 commits behind, 38 ahead - one of the ahead commits was merge. Now 37 ahead

nikitamasych committed:
:bug: imports removed from bn254 tests with cargo fix and added cfg(test) attribute

ZamDimon committed:
:adhesive_bandage: small adjustments in torus

ZamDimon committed:
:green_heart: fix conflicts on merge

ZamDimon committed:
:twisted_rightwards_arrows: Merge pull request #2 from `feature/bn254-algebraic-torus`

:doughnut: Algebraic Torus Tests and new optimized entrypoint

ZamDimon committed:
:truck: transfer `modexp` from `feature/modexp`

ZamDimon committed:
:truck: move `modexp` implementation from `era-boojum` here

ZamDimon commented:
Moved the `modexp` implementation from https://github.com/matter-labs/era-boojum/pull/38 here and transferred tests from #33. 

Also updated the title of this PR accordingly.

ZamDimon committed:
:green_heart: add normalizations to fix `uintx` issue

ZamDimon committed:
:1234: add `modexp` over `uint2048`

ZamDimon committed:
:construction: :test_tube: adding tests for 256-byte modexp implementation

ZamDimon committed:
:bug: :test_tube: make `modexp` tests read inputs properly

ZamDimon committed:
:sparkles: add support of 8-byte `modexp` exponent

nikitamasych committed:
:sparkles: added modexp entrypoint

nikitamasych committed:
:sparkles: added implementation for inner entrypoint of modexp

nikitamasych committed:
:sparkles: added modexp to scheduler, minor code quality enhancements

ZamDimon committed:
:zap: more optimizations, performance debugging tests

jules commented on src/bn254/ec_pairing/implementation.rs:
```diff
@@ -0,0 +1,747 @@
+{
+    p.enforce_reduced(cs);
+    q.enforce_reduced(cs);
+
```
it would also be useful here to do a subgroup check on p, and to include some tests for it specifically

jules commented on src/bn254/ec_pairing/implementation.rs:
```diff
@@ -0,0 +1,747 @@
+{
+    p.enforce_reduced(cs);
+    q.enforce_reduced(cs);
+
```
ditto

jules commented on src/bn254/validation.rs:
```diff
@@ -0,0 +1,143 @@
+    "19485874751759354771024239261021720505790618469301721065564631296452457478373";
+const B_TWIST_C1: &str =
+    "266929791119991161246907387137283842545076965332900288569378510910307636690";
+
```
if they dont exist already, it would be helpful to include tests for all the functions in this file just to help test coverage and have a more robust pipeline

jules commented on src/bn254/tests/ec_pairing.rs:
```diff
@@ -0,0 +1,381 @@
+            println!("EC pairing test {} has passed!", i);
+        }
+    }
+}
```
might also be useful to include a test which ensures the bilinearity property holds with the created gadgets

ZamDimon committed:
:sparkles: :memo: make less normalization calls and include performance metrics

jules commented on src/bn254/ec_pairing/implementation.rs:
```diff
@@ -0,0 +1,655 @@
+use std::sync::Arc;
```
with the fix merged for nonnative field arithmetic, we should try to normalize really only when necessary - so let's try to remove as many of those calls as possible (also in files like tower extension arithmetic and such) as it may shave constraints off the circuit

ZamDimon committed:
:test_tube: add debug performance for `modexp`

jules commented on src/bn254/ec_pairing/implementation.rs:
```diff
@@ -0,0 +1,655 @@
+
+    let mut miller_loop = MillerLoopEvaluation::evaluate(cs, p, q);
+    let final_exp = FinalExpEvaluation::evaluate_without_torus(cs, &mut miller_loop.accumulated_f);
+    final_exp.resultant_f
```
another thing - it will be useful for the separate steps to return exceptions that we can check at the end here, just to ensure that the steps were performed correctly. many other circuits in this repo do it and i think konstantins implementation will give you an idea of how to do it

jules committed:
remove superfluous normalizations, add test for torus circuit

jules committed:
cleanup

ZamDimon committed:
:test_tube: add bilinearity test

ZamDimon committed:
:zap: add more `hard_exp` implementation versions, small polishes

ZamDimon committed:
:recycle: :test_tube: add more tests for `modexp`

ZamDimon committed:
:memo: add additional documentation for `modexp`

ZamDimon committed:
:test_tube: :sparkles: make tests more convenient, add non-subgroup points test

ZamDimon committed:
:adhesive_bandage: make `hard_part` of ecpairing generalized

ZamDimon committed:
:art: run `cargo fmt`

ZamDimon committed:
:closed_lock_with_key: add additional validation in `ecpairing`

ZamDimon committed:
:closed_lock_with_key: :bug: fix curve validation function and add tests

ZamDimon commented:
@jules Added more validations and tests, as requested, and cleared up the hard exponentiation part. Could you review the changes?

jules reviewed (approved):
I think we're there. Last point, did we get any good speedups by updating the division on the torus? Or has this not been explored yet?

ZamDimon commented:
> I think we're there. Last point, did we get any good speedups by updating the division on the torus? Or has this not been explored yet?

Unfortunately, we have not received any optimizations yet. Seems we might get some performance boost by not calculating $c \gets a/b$, but merely asserting $b \cdot c == a$ inside R1CS (as the paper suggests), but I couldn't figure out how to do that inside this constraint system

nikitamasych committed:
:recycle: change modexp to 256 32 256 version

nikitamasych committed:
:see_no_evil: added DS_Store to .gitignore

nikitamasych committed:
:pushpin: pin to git version instead of the local path in Cargo.toml

ZamDimon committed:
:sparkles: make `ec_pairing` more optimized

ZamDimon committed:
:sparkles: reduce number of rows for pairing further

ZamDimon committed:
:sparkles: small polishes

ZamDimon committed:
:twisted_rightwards_arrows: Merge pull request #3: Algebraic Torus Fix

⚡ Fix Algebraic Torus
